### PR TITLE
Friendship v1

### PR DESCRIPTION
**Pull request by jacobrees**
This pull request is for the creation of a model "Friendship", the purpose of this model is to set up associations between Users and Friends.

### CODE REVIEW

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle: 

Hello, 
Thanks for the job done so far
However, there are some missing requirements

1. Kindly enable Github actions (linter) in your repository.

2. Please, add unit and integration tests as required [here](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/tests.png)

3. Running `rails db:migrate` failed
```rb
...
== 20200204165841 RemoveForeignKeys: migrating ================================
-- remove_foreign_key(:comments, :users)
   -> 0.0058s
-- remove_foreign_key(:likes, :posts)
   -> 0.0050s
-- remove_foreign_key(:likes, :users)
   -> 0.0046s
-- add_index(:comments, :user_id)
   -> 0.0350s
-- add_index(:likes, :post_id)
   -> 0.0447s
-- add_index(:likes, :user_id)
   -> 0.0452s
== 20200204165841 RemoveForeignKeys: migrated (0.1408s) =======================

== 20201006125641 CreateFriendships: migrating ================================
-- create_table(:friendships)
rails aborted!
StandardError: An error has occurred, this and all later migrations canceled:

undefined method `bolean' for #<ActiveRecord::ConnectionAdapters::PostgreSQL::TableDefinition:0x00007fbd10134370>
Did you mean?  boolean
/home/maelfosso/Documents/Microverse/TSE/RoR/ror-social-scaffold/db/migrate/20201006125641_create_friendships.rb:6:in `block in change'
/home/maelfosso/.rvm/gems/ruby-2.7.0/gems/activerecord-5.2.4.1/lib/active_record/connection_adapters/abstract/schema_statements.rb:305:in `create_table'
/home/maelfosso/.rvm/gems/ruby-2.7.0/gems/activerecord-5.2.4.1/lib/active_record/migration.rb:871:in `block in method_missing'
/home/maelfosso/.rvm/gems/ruby-2.7.0/gems/activerecord-5.2.4.1/lib/active_record/migration.rb:840:in `block in say_with_time'
/home/maelfosso/.rvm/gems/ruby-2.7.0/gems/activerecord-5.2.4.1/lib/active_record/migration.rb:840:in `say_with_time'
...
```
Fix that, please.

4. It's required, as indicated [here](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png), that posts timeline includes posts written by logged-in user and their friends. It's not that behavior implemented in your web application, kindly fix that, please.

5. another behavior is the presence of a button to invite user to friendship on users list page and on single user page. None of those two pages contains such a button. Kindly add it on both pages, please

**The Friendship is the mandatory feature of this milestone but there is no button for requesting a friendship.**

Kindly make the suggested changes and submit a new code review request
Thanks

Mael FOSSO

**Review by rOluochKe — CHANGES_REQUESTED** [review]
Hello Team
### You have done well so far :clap: However, you have a few issues to fix:-
- [x] User can see friendships requests
- [ ] User has to accept friendship request to start friendship
- [ ] User can reject a friendship request
- [x] There is a button to invite user to friendship on users list page and on single user page
### Make the changes and submit for another code review. Good luck.
- Happy coding!

**Review by trekab — CHANGES_REQUESTED** [review]
### Changes Required! :recycle: 
Hi team! Well done implementing the suggested changes from the earlier reviews. You are almost done with this milestone. There are a few vital changes to make to have this project approved. These include:
- [ ] The Posts timeline should include posts written by logged-in user and their friends ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
- [ ] Great job adding some validation tests. Consider adding some integration tests too. [Here](https://www.codewithjason.com/rails-integration-tests-rspec-capybara/) is a great article about writing integration tests. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/tests.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
- [ ] __[Optional]__ It would be nice if the friendship decline button was next to the friendship acceptance button rather than creating a different user in the view just to have the decline button.

__Kindly make the necessary changes and submit another code review request.__
_Feel free to reach me on slack for any inquiries about this review._ :wink: 
__Treasure Kabareebe__
[GitHub](https://github.com/trekab) | [LinkedIn](https://www.linkedin.com/in/treasure-kabareebe/) | [Twitter](https://twitter.com/TKabareebe)

File: app/views/posts/index.html.erb [hunk 1]
```diff
@@ -9,7 +9,35 @@
...
   </h3>
-  <ul class="posts">
-    <%= render @timeline_posts %>
+
+
+  <ul class="posts">  
+    <% @timeline_posts.each do |post|%>
+    <% if post.user.id == current_user.id  || Friendship.where(friend_id: post.user.id, confirmed: true).count > 0 || Friendship.where(user_id: post.user.id, confirmed: true).count > 0 %>
+<li class="post">
+  <div>
```
Comment (bertil291utn) [inline app/views/posts/index.html.erb:18]: ## Viewers and helpers 
- There should be NO application logic in the views nor helpers (no `Post.new`, `Comment.create`, `current_user.friends.find_by(...)`...). Everything connected with data processing should be encapsulated in models and used in controllers.
  - [ ] Move to the controller `where` property and send a new variable from there

File: app/views/users/index.html.erb [hunk 2]
```diff
@@ -1,5 +1,38 @@
 <section class="users-section">
+<h2>Friend Requests</h2>
   <ul class="users-list">
-    <%= render @users %>
+      <% @users.each do |user| %>
+        <% if user.friendships.where(friend_id: current_user.id, user_id: user.id, confirmed: false).count > 0 %>
+          <li> Friend Request: <%= user.name%> </li>
+          <%= button_to "Accept Friend Request", friendship_path(user), :method => :put %>
```
Comment (bertil291utn) [inline app/views/users/index.html.erb:7]: ## Viewers and helpers 
- There should be NO application logic in the views nor helpers (no `Post.new`, `Comment.create`, `current_user.friends.find_by(...)`...). Everything connected with data processing should be encapsulated in models and used in controllers, [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright
  - [ ] Move to the controller `where` property and send a new variable from there

File: app/views/users/index.html.erb [hunk 3]
```diff
@@ -1,5 +1,38 @@
...
+    <% if current_user.friendships.where(friend_id: user.id, user_id: current_user.id, confirmed: false).count > 0 %>   
+    <p>Friendship Request Pending</p>
+    <%= button_to "Cancel Request", friendship_path(user.id), method: :delete %>
+    <% elsif current_user.friendships.where(user_id: current_user.id, friend_id: user.id, confirmed: true).count > 0%>
+      <%= button_to "Unfriend", friendship_path(user.id), method: :delete %>
+    <% elsif user.friendships.where(user_id: user.id, friend_id: current_user.id, confirmed: true).count > 0%>
+      <%= button_to "Unfriend", friendship_path(current_user.id), method: :delete %>     
+    <% elsif user.friendships.where(user_id: user.id, friend_id: current_user.id, confirmed: false).count > 0%>
+      <%= button_to "Decline Request", friendship_path(current_user.id), method: :delete %>
+     <%= button_to "Accept Friend Request", friendship_path(user), :method => :put %>
```
Comment (bertil291utn) [inline app/views/users/index.html.erb:28]: ## Viewers and helpers 
- There should be NO application logic in the views nor helpers (no `Post.new`, `Comment.create`, `current_user.friends.find_by(...)`...). Everything connected with data processing should be encapsulated in models and used in controllers, [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright
  - [ ] Move to the controller `where` property and send a new variable from there

File: app/views/users/show.html.erb [hunk 4]
```diff
@@ -1,5 +1,21 @@
...
+    <% if current_user.friendships.where(friend_id: @user.id, user_id: current_user.id, confirmed: false).count > 0 %>   
+    <p>Friendship Request Pending</p>
+    <%= button_to "Cancel Request", friendship_path(@user.id), method: :delete %>
+    <% elsif current_user.friendships.where(user_id: current_user.id, friend_id: @user.id, confirmed: true).count > 0%>
+      <%= button_to "Unfriend", friendship_path(@user.id), method: :delete %>
+    <% elsif @user.friendships.where(user_id: @user.id, friend_id: current_user.id, confirmed: true).count > 0%>
+      <%= button_to "Unfriend", friendship_path(current_user.id), method: :delete %>     
+    <% elsif @user.friendships.where(user_id: @user.id, friend_id: current_user.id, confirmed: false).count > 0%>
+      <%= button_to "Decline Request", friendship_path(current_user.id), method: :delete %>
+     <%= button_to "Accept Friend Request", friendship_path(@user), :method => :put %>
```
Comment (bertil291utn) [inline app/views/users/show.html.erb:13]: ## Viewers and helpers 
- There should be NO application logic in the views nor helpers (no `Post.new`, `Comment.create`, `current_user.friends.find_by(...)`...). Everything connected with data processing should be encapsulated in models and used in controllers, [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright
  - [ ] Move to the controller `where` property and send a new variable from there

**Review by bertil291utn — CHANGES_REQUESTED** [review]
# **Status:** Changes requested 🟠 
<p align="center"> image <br><strong>Not yet</strong> </p>

# Hey guys, you've been doing a great job, and also I hope you enjoyed coding.

### Your work is amazing you master it Git Flow you have a professional README file. 

However, in order to accomplish all the requirements, you required to do these changes, listed below:
### MANDATORY
- Posts timeline includes posts written by logged-in user and their friends ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

  - [x] Add also the number of likes and comments  
  - [ ] The most recent comment should be on top 
  - [ ] Timeline (comments and likes) should display just for friends, this [timeline](https://user-images.githubusercontent.com/24902525/96151830-7b07e200-0ed1-11eb-8677-3308e20f1f73.png) appears when I create a new user 


### OPTIONAL
- [x] Add the username in the navbar to know which user is logged 
- [ ] Display only [once](https://user-images.githubusercontent.com/24902525/96152139-d4701100-0ed1-11eb-9836-0abb8edb25d0.png) the friendship sender name on the invitation person receiver page 

### Summary

-  No application logic on views
- Display timeline only for friends 
-  Display number of comments and dislikes 
- Most recent comments should be on top 

***

Guys work on the requested changes above and request another review.
Good luck. 

> “Never stop, never conform until what is good becomes even better, and what’s better becomes excellent.”

**Comment by Dandush03** [comment]
# Status: :red_circle: Require Changes :red_circle: 
### Great Work, but still we need some improvements.
- [ ] __User View:__ Make sure there is no logic in them, maybe [this reading might help you](http://codefol.io/posts/Where-Do-I-Put-My-Code)

# Happy Coding :muscle: :nerd_face:

**Review by voscarmv — APPROVED** [review]
# Project status: Approved ✨ 😎👍 💯 ✔️

- Great work applying all the required changes!

### You may merge this branch and move forward ⛙🛣

### Questions or comments? Reach me on [Twitter](https://twitter.com/voscarmv) or [LinkedIn](https://linkedin.com/in/oscar-mier) 🧑‍💻🗨️🧑‍💻 
# Happy Hacking! 🧠 ⌨️ 🖱️ 🖥️

**Pull request merged**